R_notebooks/06_Reshaping_data.Rmd

```diff
@@ -0,0 +1,186 @@
+---
+title: "Reshaping data with `tidyr`"
+output: html_notebook
+---
+
+## Outline
+
+-   Converting to/from tidy format
+
+## Introduction
+
+There are often cases where you'd like to change the shape of your data, by
+which we mean to change the structure of the columns and/or rows. The `tidyr`
+package contains functions to support this.
+
+As an example dataset for reshaping, we'll use data on territorial fossil CO2
+emissions by country (Friedlingstein et al. 2023)[^1].
+
+**Note:** To read the data, we'll use a call to the `read_excel` function from
+the `readxl` package. We only touched on this in notebook 2, so don't worry
+about understanding exactly how the data is read in. Treat this as a glimpse of
+seeing the Tidyverse in action 'in the wild'. You can read up about `read_excel`
+as extra homework, if you're keen :-)
+
+```{r}
+emissions <- readxl::read_excel("data/National_Fossil_Carbon_Emissions_2023v1.0.xlsx",
+                          sheet = "Territorial Emissions",
+                          skip = 11) |>
+  dplyr::rename(Year = `...1`) |>  # missing column name in data got assigned name `...1` when reading
+  dplyr::select(Year, World, Africa, Asia, `Central America`, Europe,
+                `Middle East`, `North America`, Oceania, `South America`,
+                Bunkers) |>
+  dplyr::mutate(Year = as.integer(Year))
+
+print(emissions)
+```
+
+## Converting to/from tidy format
+
+One scenario for reshaping data is when we receive data that is not in tidy
+format and want to transform it so that it is tidy.
+
+In the emissions data above, notice how each column other than 'Year'
+corresponds to a region of the world (or the category 'Bunkers'):
+
+```{r}
+colnames(emissions)
+```
+
+This is not in tidy format. The observations correspond to year-region pairs,
+while the variable being measured is territorial emissions. But the data is
+structured as a matrix, where to get the emissions value for a region and year
+you look up the entry for the appropriate row and column.
+
+What we'd like to do is create a dataframe with columns 'Year', 'Country' and
+'Territorial_emissions' columns, with rows corresponding to year-region pairs.
+
+### Matrix format to tidy format with `pivot_longer`
+
+We can use the function `pivot_longer` from the `tidyr` package to help us make
+the required transformation. Like most of the Tidyverse functions, there are a
+lot of possible parameters to use in this function, but we only need to use it
+as follows:
+
+```         
+tidyr::pivot_longer(data, <columns-to-gather>, names_to = "Name_for_gathered_columns", values_to = "Variable_name")
+```
+
+where
+
+-   `data` is the dataframe to transform.
+-   `<columns-to-gather>` is a specification of the columns that should be
+    gathered into a new column.
+-   `names_to` defines the name of the new column to create, whose values will
+    be the names in `<columns-to-tidy>`.
```

Comment:
Should "columns-to-tidy" be renamed to "Name_for_gathered_columns" or vice versa - to match with line 66?

Reply:
In fact, it should actually be changed to `<columns-to-gather>`. Thanks for spotting that.